### PR TITLE
Implement Spool Interactions

### PR DESCRIPTION
**Pull request by CodyWallbridge**
changelog:
- Added All Spools and Joined Spools section to sidebar
- Added Join/Leave/Manage spool button to the Spools top section
- Added Rules section to spools info section
- Added manage spool page
- Added functionality to add and remove mods (they cannot do anything yet as permissions are not implemented) (takes username)
- Added functionality to change the owner (takes username)
- Added tests to ensure 100% code coverage in backend tests
- Added error throwing to backend (if you try to change ownership or add mod for a user which does not exist, error shows in front end)
- Added delete spool functionality with delete confirmation
- Added delete user in case it is needed
- Ensured cascade works when deleting a spool (throws error)
- Ensured that mods exist when you create a spool (throws error)

Pictures:
image
image
image
image
image

### CODE REVIEW

**Comment by nathanpoppe** [comment]
Functionality seems to be mostly working but I found a few issues:
- The remove moderator "Delete" button does not seem to work
- The owner can add themself as a moderator, is this expected?
- I think attempting to change the owner to the current owner should do nothing
- Adding a duplicate moderator should have a different error message
- The Save button is a bit misleading because as far as I can tell it only saves the rules?

Some potential UI ideas:
- In the spool view, should we move the buttons (Create Post, Manage Spool) and rules to another column beside the threads (similar to Reddit), not a huge deal could be done later on
- In the Spool view, I think the rules should be a text element instead of a disabled text area
- Should the rules section just be a more general Spool description instead?
- In the Spool Settings, should add Padding above "Add" and "Change" buttons and error messages
- Wondering if it would be a good idea to change the moderator sections to one list where you can add, remove, change to owner, etc. from. Something like this?

image

Overall, nice work, you added a ton!

**Comment by CodyWallbridge** [comment]
> Functionality seems to be mostly working but I found a few issues:
> 
> * The remove moderator "Delete" button does not seem to work
> * The owner can add themself as a moderator, is this expected?
> * I think attempting to change the owner to the current owner should do nothing
> * Adding a duplicate moderator should have a different error message
> * The Save button is a bit misleading because as far as I can tell it only saves the rules?
> 
> Some potential UI ideas:
> 
> * In the spool view, should we move the buttons (Create Post, Manage Spool) and rules to another column beside the threads (similar to Reddit), not a huge deal could be done later on
> * In the Spool view, I think the rules should be a text element instead of a disabled text area
> * Should the rules section just be a more general Spool description instead?
> * In the Spool Settings, should add Padding above "Add" and "Change" buttons and error messages
> * Wondering if it would be a good idea to change the moderator sections to one list where you can add, remove, change to owner, etc. from. Something like this?
> 
> image
> 
> Overall, nice work, you added a ton!

Awesome points. I agree with all of the first points, I will implement them while double checking code coverage.
Saving does only save rules because the rest happens dynamically. I could update the button to say "Save Rules" and move it up to the rules section? I thought there would be more things to edit but it just didnt work out that way.

text vs disabled text area for rules: I tried. Text will not show new line characters. So even if rules are separated by new lines to make a list, text puts it all on one line which looks terrible. I think the current implementation sucks as well, but could not find a way to make it better. If we have time and someone really wants to, I definitely agree with changing it up. I just dont know how to. Though I do agree with changing it to like a Rules and Description area. I can work on adding that. Unfortunately its also extremely difficult to get the text area to flex fit all the contents. so it doesnt auto expand to fit all the rows which is very annoying and I cant find a simple way to fix this. For now, its functional. Also something to be looked at later.

I'm impartial to moving the stuff to another sidebar. up to the group. im ok either way

I agree padding between add/change and errors. will do that later

The list you linked looks super nice, but not sure how implementing that would go to be honest lol. maybe another idea for later if we have time?

**Comment by noahc3** [comment]
Functionality comments/ideas:

Spool page:
- Spool rules box should be moved outside of the isAuthenticated check -- it should show whether you are logged in or not.
- Can we add padding to the VStack for the buttons bar and rules so its the same width as the post boxes?
  - Or move these to a sidebar (now or later, but I'd personally prefer this eventually)
- Instead of textarea, use Text for spool rules. Add `whiteSpace="pre-wrap"` property to show newlines properly.

Spool settings page:
- Small margin between textbox inputs and the Add/Change buttons would look nice
  - Or, these buttons can be inside the text input like [this example](https://chakra-ui.com/docs/components/input#password-input-example)
  example:
image
- Hide "remove moderators" header and horizontal rule if there are no users to display?
-  Idea: Move "Save Rules" button up above under Rules/Description textbox and make it not navigate back to spool. Replace existing "Save Rules" button with just a "Back" button to return to spool page.
- Idea: make moderator delete buttons `size='sm'`, perhaps even remove the delete text? (So then just the trash icon, can use `IconButton` instead of `Button`)
  example:
image

File: threadit-api/Controllers/v1/SpoolController.cs [hunk 1]
```diff
@@ -59,11 +59,76 @@ public async Task<IActionResult> GetAllSpools([FromServices] SpoolService spoolS
...
+            catch (Exception e)
+            {
+                return BadRequest(e.Message);
+            }
+
+            return Ok(spool);
+        }
+
+        [HttpGet("change/{spoolId}/{userName}")]
+        public async Task<IActionResult> ChangeOwner([FromRoute] string spoolId, [FromRoute] string userName, [FromServices] SpoolService spoolService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/SpoolController.cs:99]: should have same auth check as above

File: threadit-api/Controllers/v1/SpoolController.cs [hunk 2]
```diff
@@ -59,11 +59,76 @@ public async Task<IActionResult> GetAllSpools([FromServices] SpoolService spoolS
             return Ok(spools);
         }
 
+        [HttpGet("delete/{spoolId}")]
+        public async Task<IActionResult> DeleteSpool([FromRoute] string spoolId, [FromServices] SpoolService spoolService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/SpoolController.cs:63]: - add `[AuthenticationRequired]` attribute
- check and make sure calling user ID (`request.GetUser()`) matches the current owner before changing

File: threadit-api/Controllers/v1/SpoolController.cs [hunk 3]
```diff
@@ -59,11 +59,76 @@ public async Task<IActionResult> GetAllSpools([FromServices] SpoolService spoolS
...
+
+        [HttpGet("mods/{spoolId}")]
+        public async Task<IActionResult> AllModsForSpool([FromRoute] string spoolId, [FromServices] SpoolService spoolService)
+        {
+            UserDTO[]? users = await spoolService.GetAllModsForSpoolAsync(spoolId);
+            return Ok(users);
+        }
+
+        [HttpGet("mods/add/{spoolId}/{userName}")]
+        public async Task<IActionResult> AddModerator([FromRoute] string spoolId, [FromRoute] string userName, [FromServices] SpoolService spoolService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/SpoolController.cs:84]: should have same auth check as above

File: threadit-api/Controllers/v1/SpoolController.cs [hunk 4]
```diff
@@ -59,11 +59,76 @@ public async Task<IActionResult> GetAllSpools([FromServices] SpoolService spoolS
...
+            }
+            catch (Exception ex)
+            {
+                return BadRequest(ex.Message);
+            }
+            return Ok(spool);
+        }
+
+        [HttpGet("mods/remove/{spoolId}/{userId}")]
+        public async Task<IActionResult> RemoveModerator([FromRoute] string spoolId, [FromRoute] string userId, [FromServices] SpoolService spoolService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/SpoolController.cs:114]: should have same auth check as above

File: threadit-api/Controllers/v1/SpoolController.cs [hunk 5]
```diff
@@ -59,11 +59,76 @@ public async Task<IActionResult> GetAllSpools([FromServices] SpoolService spoolS
...
+        {
+            Spool? spool = await spoolService.RemoveModeratorAsync(spoolId, userId);
+            return Ok(spool);
+        }
+
+        [HttpPost("save/{spoolId}")]
+        public async Task<IActionResult> SaveRules([FromRoute] string spoolId, [FromBody] SaveRulesRequest rules, [FromServices] SpoolService spoolService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/SpoolController.cs:121]: should have same auth check as above

File: threadit-api/Controllers/v1/UserSettingsController.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace ThreaditAPI.Controllers.v1
+{
+    [ApiController]
+    [Route("v1/userSettings")]
+    public class UserSettingsController : ControllerBase
+    {
+
+        [HttpGet("remove/{userId}/{spoolName}")]
+        public async Task<IActionResult> RemoveSpoolUser([FromRoute] string userId, [FromRoute] string spoolName, [FromServices] UserSettingsService userSettingsService)
```
Comment (noahc3) [inline threadit-api/Controllers/v1/UserSettingsController.cs:13]: - add `[AuthenticationRequired]` attribute
- remove {userId} from path and just get the userId from `request.GetUser()`

File: threadit-api/Controllers/v1/UserSettingsController.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            UserSettings userSettings = await userSettingsService.RemoveUserSettingsAsync(userId, spoolName);
+            return Ok(userSettings);
+        }
+
+        [HttpGet("join/{userId}/{spoolName}")]
```
Comment (noahc3) [inline threadit-api/Controllers/v1/UserSettingsController.cs:19]: same as remove spool user

File: threadit-api/Repositories/SpoolRepository.cs [hunk 8]
```diff
@@ -44,35 +69,67 @@ public async Task InsertSpoolAsync(Spool spool)
...
-                return dbSpool.Moderators.First().Split(',').ToList();
+                if (!dbSpool.Moderators.IsNullOrEmpty())
+                {
+                    foreach (string currentId in dbSpool.Moderators)
+                    {
+                        UserDTO? currentUser = await db.Users.FirstOrDefaultAsync(u => u.Id == currentId);
+                        if (currentUser != null)
+                        {
+                            Array.Resize(ref users, users.Length + 1);
+                            users[^1] = currentUser;
```
Comment (noahc3) [inline threadit-api/Repositories/SpoolRepository.cs:86]: god i love that microsoft added the array ranges and indexing from python

File: threadit-web/src/api/SpoolAPI.ts [hunk 9]
```diff
@@ -70,4 +77,92 @@ export default class SpoolAPI {
...
+
+        if (!response.ok) {
+            throw new Error(`Failed to get all mods for the spool: ${await response.text()}`);
+        }
+
+        return await response.json();
+    }
+
+    static async removeModerator(spoolId: string, userId: string): Promise<IUserProfile> {
+        const response = await get(removeModeratorEndpoint + spoolId + '/' + userId);
```
Comment (noahc3) [inline threadit-web/src/api/SpoolAPI.ts:92]: if you add authentication to these endpoints you will need to change `get` to `getWithAuth` (for most of these)

File: threadit-web/src/api/SpoolAPI.ts [hunk 10]
```diff
@@ -70,4 +77,92 @@ export default class SpoolAPI {
...
+        const response = await get(addModeratorEndpoint + spoolId + '/' + userName);
+        if (response.ok) {
+            //returns 1 if the add was successful
+            return 1;
+        }
+
+        else {
+            //otherwise change failed and need to see why
+            const errorText = await response.text();
+            if (errorText === "User does not exist.") {
```
Comment (noahc3) [inline threadit-web/src/api/SpoolAPI.ts:111]: why not just return the text? and `true` if it succeeded.

return type is then `Promise<bool | string>`

that way you dont need your if statements when you display the error later

Comment (CodyWallbridge) [reply threadit-web/src/api/SpoolAPI.ts:111]: because I didnt know you could do that lol

**Review by noahc3 — APPROVED** [review]

**Review by nathanpoppe — APPROVED** [review]

**Pull request merged**